ui: migrate to nodes api

psychedelicious opened the pull request:
1. `yarn install`
2. Start nodes web server
```bash
# from the repo root
python scripts/invoke-new.py --web
```
3. `yarn dev`
4. click the Nodes tab, create then invoke

see [README `invokeai/frontend/web/src/services/api/README.md`](https://github.com/psychedelicious/InvokeAI/blob/feat/ui/api-migration/invokeai/frontend/web/src/services/api/README.md) for info on the client generation

maryhipp commented:
I think it would be helpful if we had some sort of annotation for which of these files are generated automatically

psychedelicious commented:
Easiest thing is to put them all in a folder - everything in `src/services/api` except the readme and python script are generated. Feel free to organize

Last night I didn't have much time, just made the readme and cleaned up a lot of random stuff from my experiments with generators. 

Next step is to make thunks of the calls via `createAsyncThunk` and write a function to transform the redux state into a valid graph.

psychedelicious commented:
You can see some past work in this branch, which parsed the schema at runtime and generated a node editor UI using `react-flow`: https://github.com/psychedelicious/InvokeAI/tree/development-invoke/frontend/src/features/nodeEditor

It won't be useful until we are working on the node editor UI, and it isn't functional now (it was and worked great back when I wrote it! but the underlying nodes server has changed and this `development-invoke` branch now represents a failed attempt at rebasing an even older node editor branch)

I'm sad to say that my screen recordings of this have been deleted, but it was pretty cool.

psychedelicious commented:
tonight's updates:
- all generated code is in `src/services/api/`
- temp redux apiSlice for getting things bootstrapped and figuring out how to fit the pieces together
- added examples of `createAsyncThunk`
- added progress images (see #2959 for merging changes into main)
- added comments along the way

maryhipp commented on invokeai/frontend/web/src/app/NodeAPITest.tsx:
```diff
@@ -0,0 +1,200 @@
+    // TODO: write types for the socket.io payloads, haven't found a generator for them yet...
+
+    // subscribe to the current session
+    socket.emit('subscribe', { session: sessionId });
```
Where is the listener for this on the backend?

maryhipp replied:
So each time we hit the REST API to create a session, we will also initialize this socket subscriber for each one? And we have a separate connection (room? still learning the socket lingo) for the global state like server connection? Is that why there are two socket proxies in the vite build file now?

psychedelicious replied:
> Where is the listener for this on the backend?

Listener is here: https://github.com/psychedelicious/InvokeAI/blob/f02be7ed34798307a8000b4ad9ecc75841f64a32/invokeai/app/api/sockets.py

psychedelicious replied:
> So each time we hit the REST API to create a session, we will also initialize this socket subscriber for each one? And we have a separate connection (room? still learning the socket lingo) for the global state like server connection? Is that why there are two socket proxies in the vite build file now?

This is probably the wrong approach in the long term, but yes in the little demo the client is entering a new room, generating, then leaving when finished generating.

There is still a client connection that is established during that initial HTTP request to the socketio server. socketio identifies connectio with a client `sid` - session id - no relation to nodes "sessions", except that the node session id is exactly that `sid` (😅).

In the dev tools, we are looking at that overall socketio session, and then within that we may subscribe to any number of rooms.

The two proxies in the vite config proxy requests from the vite dev server (on port 5173) to the actual socketio server. One of the proxy setups is for the old server, and the second proxy is for the nodes server. 

The proxy is for the overall socketio connection; the room subscriptions, progress events, etc are all happening within that connection

maryhipp commented on invokeai/frontend/web/src/app/NodeAPITest.tsx:
```diff
@@ -0,0 +1,200 @@
+    });
+
+    // not sure when we get this?
+    socket.on('session_complete', (data) => {
```
I asked kyle the difference between `invocation_complete` and `graph_execution_state_complete`. 

> invocation is for a single node (I think I need to go through the code and rename them all to use "node"). Session is for the entire graph completing execution (at least everything that's in the graph).

psychedelicious commented:
I've made a branch on the `invoke-ai/InvokeAI` repo to work from:
https://github.com/invoke-ai/InvokeAI/tree/feat/ui/nodes-api

Closing this PR in favour of working from a branch.
